[WIP][SPARK-31198][CORE] Use graceful decommissioning as part of dynamic scaling

holdenk opened the pull request:
This is WIP since it is on top of SPARK-31197 (which itself is WIP on top off SPARK-20629 ) and should probably have more testing. We use SPARK-31197's exiting of the executor once decommissioning is finished to allow us to replace the usage of killExecutor with decommission executor when enabled during dynamic allocation.

### What changes were proposed in this pull request?

If graceful decommissioning is enabled, Spark's dynamic scaling uses this instead of directly killing executors.

### Why are the changes needed?

When scaling down Spark we should avoid triggering recomputes as much as possible.

### Does this PR introduce _any_ user-facing change?

Hopefully their jobs run faster. It also enables experimental shuffle service free decommissioning when graceful decommissioning is enabled.

### How was this patch tested?

For now I've extended the ExecutorAllocationManagerSuite to cover this.

SparkQA commented:
**[Test build #123956 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123956/testReport)** for PR 28818 at commit [`2ff94ec`](https://github.com/apache/spark/commit/2ff94ece8eb9b33524fd6705bfcaca6427855450).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28578/
Test PASSed.

SparkQA commented:
**[Test build #123956 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123956/testReport)** for PR 28818 at commit [`2ff94ec`](https://github.com/apache/spark/commit/2ff94ece8eb9b33524fd6705bfcaca6427855450).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123956/
Test FAILed.

agrawaldevesh reviewed (commented):
Hi @holdenk .. Using proper decommissioning for dynamic allocation would be great and help unify these codepaths. Thank you again for working on this. 

My two comments below are mere nits and just to make sure I am following along. It reads fine as is.

holdenk commented on core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala:
```diff
@@ -333,11 +335,19 @@ private[spark] class ExecutorMonitor(
       UNKNOWN_RESOURCE_PROFILE_ID)
+
+    // Check if it is a shuffle file, or RDD to pick the correct codepath for update
+    if (event.blockUpdatedInfo.blockId.isInstanceOf[ShuffleDataBlockId] && shuffleTrackingEnabled) {
+      event.blockUpdatedInfo.blockId match {
+        case ShuffleDataBlockId(shuffleId, _, _) => exec.addShuffle(shuffleId)
+        case _ => // For now we only update on data blocks
+      }
```
So it's not (I still want to get to SPARK-31974). The executor monitor keeps track of locations of cache and shuffle blocks and this can be used to decide which executor(s) Spark should shutdown first. Since we move shuffle blocks around now this wires it up so that it keeps track of it.

agrawaldevesh replied:
Makes perfect sense ! A comment in the code would be helpful ;-).

holdenk replied:
Sure, I'll add one :)

holdenk commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -497,6 +453,70 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
 
+  /**
+   * Request that the cluster manager decommission the specified executors.
+   * Default implementation delegates to kill, scheduler must override
+   * if it supports graceful decommissioning.
```
So that's copied from the trait it's implementing (it is the part doing the override), I'll delete it here though since it's confusing in this context, good catch.

SparkQA commented:
**[Test build #123995 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123995/testReport)** for PR 28818 at commit [`ef3f523`](https://github.com/apache/spark/commit/ef3f52364afffb9943d0e0a9e36e2b043c4a3284).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28617/
Test PASSed.

SparkQA commented:
**[Test build #123995 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123995/testReport)** for PR 28818 at commit [`ef3f523`](https://github.com/apache/spark/commit/ef3f52364afffb9943d0e0a9e36e2b043c4a3284).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123995/
Test FAILed.

SparkQA commented:
**[Test build #124139 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124139/testReport)** for PR 28818 at commit [`7691d2d`](https://github.com/apache/spark/commit/7691d2d31bc3677902c7d2b9b74ec976dd3342e2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28756/
Test PASSed.

SparkQA commented:
**[Test build #124139 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124139/testReport)** for PR 28818 at commit [`7691d2d`](https://github.com/apache/spark/commit/7691d2d31bc3677902c7d2b9b74ec976dd3342e2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124139/
Test FAILed.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala:
```diff
@@ -333,11 +335,26 @@ private[spark] class ExecutorMonitor(
+       * committed.
+       */
+      event.blockUpdatedInfo.blockId match {
+        case ShuffleDataBlockId(shuffleId, _, _) => exec.addShuffle(shuffleId)
```
Since we are touching `ExecutorMonitor`, when do we have a counter operation, `exec.removeShuffle`? In this PR, it seems that `executorsKilled` is used. Is it enough?
cc @dbtsai

holdenk replied:
Yeah so since we're only doing migrations during decommissioning, whatever shuffle files remain on the host when it dies will do the cleanup. I can't think of why we would need to do a delete operation here as well, but if it would be useful for your follow on work I can add it?

SparkQA commented:
**[Test build #124154 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124154/testReport)** for PR 28818 at commit [`9bb0293`](https://github.com/apache/spark/commit/9bb0293125af3dcde87b52be5bf0ba4dcc5f4a0f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28770/
Test PASSed.

SparkQA commented:
**[Test build #124154 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124154/testReport)** for PR 28818 at commit [`9bb0293`](https://github.com/apache/spark/commit/9bb0293125af3dcde87b52be5bf0ba4dcc5f4a0f).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124154/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28795/
Test PASSed.

SparkQA commented:
**[Test build #124179 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124179/testReport)** for PR 28818 at commit [`81f9dbb`](https://github.com/apache/spark/commit/81f9dbb17160e64ac7d77e3e8de2e3f0f51fabe9).

SparkQA commented:
**[Test build #124179 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124179/testReport)** for PR 28818 at commit [`81f9dbb`](https://github.com/apache/spark/commit/81f9dbb17160e64ac7d77e3e8de2e3f0f51fabe9).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124179/
Test FAILed.

SparkQA commented:
**[Test build #124181 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124181/testReport)** for PR 28818 at commit [`2674055`](https://github.com/apache/spark/commit/2674055de049c606ca32ffeb34bcc66235e14d1c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28799/
Test PASSed.

SparkQA commented:
**[Test build #124181 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124181/testReport)** for PR 28818 at commit [`2674055`](https://github.com/apache/spark/commit/2674055de049c606ca32ffeb34bcc66235e14d1c).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124181/
Test FAILed.

gatorsmile commented:
cc @Ngone51 @jiangxb1987

HyukjinKwon commented:
cc @tgravescs too

holdenk commented:
To be clear this a WIP PR and not yet ready for review. I created it to give additional context after talking with @agrawaldevesh about our respective goals. I'll try and get a "read for review" PR out next week.

holdenk commented:
Just as a follow up since @HyukjinKwon requested an SPIP for this work I won't be moving this PR out of WIP this week as originally planned.

holdenk committed:
Fix our exiting and cleanup thread for better debugging next time. Cleanup the locks we use in decommissioning and clarify some more bits.

agrawaldevesh commented on core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala:
```diff
@@ -81,6 +81,35 @@ private[spark] trait ExecutorAllocationClient {
+   */
+  def decommissionExecutor(executorId: String): Boolean = {
+    val decommissionedExecutors = decommissionExecutors(Seq(executorId),
+      adjustTargetNumExecutors = true)
```
Why is `adjustTargetNumExecutors` defaulting to true here ? This would mean that all schedulers would try to replinish the executor when asked to `DecommissionExecutor(...)` -- for example by the Master or when an executor gets a SIGPWR. 

I think it shouldn't be the default -- it should atleast be configurable. It only makes sense to have `adjustTargetNumExecutors=true` when called from `org.apache.spark.streaming.scheduler.ExecutorAllocationManager#killExecutor` (ie when it is truly called from dynamic allocation codepath and we have decided that we want to replinish the executor).

holdenk replied:
If you look above there is a configurable call. This matches how `killExecutor` is implemented down on line 124.

agrawaldevesh replied:
Can you please point me to where is the configurable call ? I don't see a config check in the code paths that call this method.

It's fine for killExecutor to unconditionally adjust the target number of executors because it is only called in the dynamic allocation codepath, but decommissionExecutor would be called from many other codepaths as well (for example when the driver gets a DecommissionExecutor message) -- and thus I think it should just assume that it should replenish the executor.

holdenk replied:
Look on line 95 of this file. I think we should match the semantics of `killExecutor` as much as possible. If there's a place where we don't want it we can use `decommissionExecutors`

agrawaldevesh replied:
Hmm, Should we rename decommissionExecutor (singular) to decommissionAndKillExecutor to reflect its purpose better ? It would be too easy to confuse it with decommissionExecutors (on line 95 of this file which allows to not replenish the target number of executors).

Do you want to make the change to the callers of decommissionExecutor in this PR and switch them to `decommissioExecutors(Seq(executorId), false)` instead. The ones I am most concerned about are:

- The handling of message DecommissionExecutor (both sync and async variants) in CoarseGrainedSchedulerBackend
- StandaloneSchedulerBackend.executorDecommissioned

In both the above cases, I think we may not _always_ want replenishing. For example, in the standalone case, when the Worker gets a SIGPWR -- do we want to replenish the executors on the remaining workers (ie oversubscribe the remaining workers) ? Similarly when an executor gets a SIGPWR, do we want to put that load on the remaining executors ? I think the answer to both should be NO unless we are doing a dynamic allocation.

Personally I am fine with any choice of naming here as long as the semantics are not silently changed under the cover, as is the case presently.

holdenk replied:
It's a new function, what are we changing?

agrawaldevesh replied:
ExecutorAllocationClient is a base class of CoarseGrainedSchedulerBackend. We moved decommissionExecutor from the latter class to the former and as such it is not a new function. Since CoarseGrainedSchedulerBackend no longer overrides decommissionExecutor, ExecutorAllocationClient.decommissionExecutor will be called when CoarseGrainedSchedulerBackend gets a DecommissionExecutor message -- and the semantics of that codepath have been changed to unconditionally impose adjustTargetNumExecutors=true.

holdenk replied:
cool, I'll update the previous calls to `decommissioExecutor`

agrawaldevesh reviewed (commented):
One question inline please.

I didn't quite follow why this PR is rebased on top of https://issues.apache.org/jira/browse/SPARK-31197 and https://issues.apache.org/jira/browse/SPARK-20629. They seem orthogonal to me. Can you please update the PR description to reflect the relationship b/w this PR and these Jira tickets. Thanks !

holdenk committed:
Shutdown executor once we are done decommissioning

Because the mock always says there is an RDD we may replicate more than once, and now that there are independent threads

holdenk committed:
Make Spark's dynamic allocation use decommissioning

holdenk committed:
Track the decommissioning executors in the core dynamic scheduler so we don't scale down too low, update the streaming ExecutorAllocationManager to also delegate to decommission

Fix up executor add for resource profile

holdenk commented:
@agrawaldevesh So this PR depends on the behaviour of the VM eventually exiting (https://issues.apache.org/jira/browse/SPARK-31197 ) since it's replacing the usage of killExecutor during dynamic allocation.

SparkQA commented:
**[Test build #126273 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126273/testReport)** for PR 28818 at commit [`9565c40`](https://github.com/apache/spark/commit/9565c4051a20f26c409f038543742161507ca593).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30889/
Test PASSed.

agrawaldevesh commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -419,48 +419,6 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
-    /**
-     * Mark a given executor as decommissioned and stop making resource offers for it.
-     */
-    private def decommissionExecutor(executorId: String): Boolean = {
```
@holdenk ... to answer your question: This block of code was moved to the base class `ExecutorAllocationClient`. So the code in `ExecutorAllocationClient` is not "New". Furthermore, the semantics of this code were changed as it was moved to now unconditionally replenish the executors.

holdenk replied:
Makes sense.

holdenk committed:
Verify executors decommissioned, then killed by external external cluster manager are re-launched

holdenk committed:
Verify some additional calls are not occuring in the executor allocation manager suite.

SparkQA commented:
**[Test build #126281 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126281/testReport)** for PR 28818 at commit [`3db60f2`](https://github.com/apache/spark/commit/3db60f24b04a278ca4e900d48ff8454c10ef0659).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30897/
Test PASSed.

holdenk committed:
Dont' close the watcher until the end of the test

holdenk committed:
Use decommissionExecutors and set adjustTargetNumExecutors to false so that we can match the pattern for killExecutor/killExecutors

SparkQA commented:
**[Test build #126273 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126273/testReport)** for PR 28818 at commit [`9565c40`](https://github.com/apache/spark/commit/9565c4051a20f26c409f038543742161507ca593).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126273/
Test PASSed.

SparkQA commented:
**[Test build #126281 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126281/testReport)** for PR 28818 at commit [`3db60f2`](https://github.com/apache/spark/commit/3db60f24b04a278ca4e900d48ff8454c10ef0659).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126281/
Test PASSed.

holdenk committed:
bump numparts up to 6

SparkQA commented:
**[Test build #126292 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126292/testReport)** for PR 28818 at commit [`daf96dd`](https://github.com/apache/spark/commit/daf96ddae8d1cc2440b8b5452a1cd7c3e499f278).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/30907/
Test PASSed.

SparkQA commented:
**[Test build #126292 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126292/testReport)** for PR 28818 at commit [`daf96dd`](https://github.com/apache/spark/commit/daf96ddae8d1cc2440b8b5452a1cd7c3e499f278).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126292/
Test FAILed.

holdenk committed:
Revert "bump numparts up to 6"

This reverts commit daf96ddae8d1cc2440b8b5452a1cd7c3e499f278.

SparkQA commented:
**[Test build #126435 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126435/testReport)** for PR 28818 at commit [`f921ddd`](https://github.com/apache/spark/commit/f921ddd3913d8b749abd2d8de645e5d098d73823).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31049/
Test PASSed.

SparkQA commented:
**[Test build #126435 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126435/testReport)** for PR 28818 at commit [`f921ddd`](https://github.com/apache/spark/commit/f921ddd3913d8b749abd2d8de645e5d098d73823).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126435/
Test FAILed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31999/
Test PASSed.

agrawaldevesh commented:
@holdenk can this PR be abandoned/closed now since this is finally in ?
